#1361 Add eo-runtime dependency even if foreign dependencies are absent

volodya-lombrozo opened the pull request:
Closes: #1361

I've added one more step in `ResolveMojo`. It adds runtime dependency if it is absent.
That solution allows build EO programs without foreing objects.

volodya-lombrozo commented:
@mximp @yegor256 pls, take a look

mximp commented:
@volodya-lombrozo can we squash commits?

mximp commented:
@volodya-lombrozo please run the build with Qulice locally.

volodya-lombrozo commented:
> @volodya-lombrozo please run the build with Qulice locally.

[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------

mximp commented on eo-maven-plugin/src/main/java/org/eolang/maven/RuntimeDependency.java:
```diff
@@ -0,0 +1,102 @@
+ *
+ * @since 0.28.11
+ */
+final class RuntimeDependency {
```
@volodya-lombrozo Since it's specific to EO runtime - I would call it `EoRuntime`

volodya-lombrozo replied:
Renamed to `EoRuntimeDependency`

mximp commented on eo-maven-plugin/src/main/java/org/eolang/maven/RuntimeDependency.java:
```diff
@@ -0,0 +1,102 @@
+     * @param other Other dependency
+     * @return True if other dependency is the eo-runtime dependency
+     */
+    boolean theSameAs(final Dependency other) {
```
@volodya-lombrozo Could be simplified to `same`

volodya-lombrozo replied:
Renamed

mximp commented on eo-maven-plugin/src/main/java/org/eolang/maven/RuntimeDependency.java:
```diff
@@ -0,0 +1,102 @@
+     *
+     * @return Maven Dependency
+     */
+    Dependency toDependency() {
```
@volodya-lombrozo Could be simplified to `dependency`

volodya-lombrozo replied:
Renamed

volodya-lombrozo commented on eo-maven-plugin/src/main/java/org/eolang/maven/ResolveMojo.java:
```diff
@@ -202,6 +203,30 @@ private Collection<Dependency> deps() throws IOException {
+    /**
+     * Check if runtime dependency is absent.
+     * @param deps Dependencies
+     *  fixme: We are using hardcoded version of RuntimeDependency.
```
I've tried to create a puzzle, but it complains that I have to link to some issue. As far as I know, I have to put current issue [#35](https://github.com/objectionary/eoc/issues/35), but we are not in `eoc` repository. `eo` repository has it's own #35 issue.

mximp replied:
> I've tried to create a puzzle, but it complains that I have to link to some issue. As far as I know, I have to put current issue [#35](https://github.com/objectionary/eoc/issues/35), but we are not in `eoc` repository. `eo` repository has it's own #35 issue.

In this case we must create new issue within `eo` repository since it is a fix for EO.

volodya-lombrozo replied:
New issue #1361

volodya-lombrozo replied:
Fixed

volodya-lombrozo commented:
> @volodya-lombrozo can we squash commits?

Done

volodya-lombrozo commented on eo-maven-plugin/src/main/java/org/eolang/maven/ResolveMojo.java:
```diff
@@ -194,6 +194,7 @@ private Collection<Dependency> deps() throws IOException {
             tojo.set(AssembleMojo.ATTR_JAR, coords);
         }
         this.checkConflicts(deps);
+        addRuntimeDependency(deps);
```
Fixed

mximp commented on eo-maven-plugin/src/main/java/org/eolang/maven/ResolveMojo.java:
```diff
@@ -202,6 +203,30 @@ private Collection<Dependency> deps() throws IOException {
+     * @param dependency Dependency
+     * @return True if dependency is runtime dependency
+     */
+    private static boolean isRuntimeDependency(final Dependency dependency) {
```
@volodya-lombrozo This can be inlined. I believe no need in separate static method.

volodya-lombrozo replied:
Done

mximp commented on eo-maven-plugin/src/main/java/org/eolang/maven/ResolveMojo.java:
```diff
@@ -202,6 +203,30 @@ private Collection<Dependency> deps() throws IOException {
+     * @return True if dependency is runtime dependency
+     */
+    private static boolean isRuntimeDependency(final Dependency dependency) {
+        return new EoRuntimeDependency().same(dependency);
```
@volodya-lombrozo May we create only one instance of `EoRuntimeDependency` and reuse it?

volodya-lombrozo replied:
Done

mximp commented:
@volodya-lombrozo looks good to me.
@yegor256 please merge (if no comments from your side).

yegor256 commented:
@volod looks very good, but I have a few questions. First, why do we need to create a new class `EoRuntimeDependency` if we already have `Dependency` class form Maven? In general, keep a simple rule in mind: if your class doesn't implement any interfaces -- you most probably are doing something wrong. It seems that you only need `equals` from this new class. Make this equals a new class, or just a new method. But `EoRuntimeDependency` seems to be a pretty procedural bag of data.

Second, `MockMavenCentral` has two public methods, while it implements an interface. This is again a mistake in OOP. The method `assertLastDownloadedDependencyAndPathEquals` (aside from the fact that the name is super ugly :) must be taken away from this class and be implemented as a new Hamcrest assertion (instance of `Assertion`), I believe.

volodya-lombrozo commented:
@yegor256 pls, take a loot at the last changes. I’ve change `MockMavenCentral` class as you suggested and removed the `EoRuntimeDependency`. 

As for `EoRuntimeDependency`, It definitely looks like a “bag with data” for now, I agree. But in the future I think we will Implement something like `HttpEoRuntimeDependency` or `ConstantEoRuntimeDependency` or `ConfigEoRuntimeDependency` - all of them will implement their own way of retrieving a eo runtime dependency from different sources. All of that classes will have one public interface method `toMavenDependency` and a common interface of cource. 

Also, it’s a quite dangerous to use `Dependency` class directly, because it has a bad architecture (not an OOP style at all) which influences a lot on overall program design (please, take a look on that [code](https://github.com/objectionary/eo/blob/21da7c8e534021f427cae8201bb8e30d7db4ae08/eo-maven-plugin/src/main/java/org/eolang/maven/ResolveMojo.java#L257)).  Moreover, we can’t change that class. I think it’s better to encapsulate it and to use something our own with a better design.

yegor256 commented on eo-maven-plugin/src/main/java/org/eolang/maven/ResolveMojo.java:
```diff
@@ -194,6 +194,7 @@ private Collection<Dependency> deps() throws IOException {
             tojo.set(AssembleMojo.ATTR_JAR, coords);
         }
         this.checkConflicts(deps);
+        ResolveMojo.addRuntimeDependency(deps);
```
@volodya-lombrozo check this out: https://www.yegor256.com/2015/12/08/temporal-coupling-between-method-calls.html

volodya-lombrozo replied:
In order to obrain the issue I had to add a few private static classes. Please, take a look at `AllDependencies` and `UniqueDependencies` classes.

yegor256 commented on eo-maven-plugin/src/main/java/org/eolang/maven/ResolveMojo.java:
```diff
@@ -202,6 +203,27 @@ private Collection<Dependency> deps() throws IOException {
+     *  or from config files.
+     */
+    private static void addRuntimeDependency(final Collection<Dependency> deps) {
+        final RuntimeDependencyEquality runtime = new RuntimeDependencyEquality();
```
@volodya-lombrozo check this out: https://www.yegor256.com/2015/09/01/redundant-variables-are-evil.html

volodya-lombrozo replied:
Fixed

yegor256 commented:
@volodya-lombrozo please, forgive me for being picky. I think that `MockMavenCentral` is over-design in this case. Here is why: The live class `Central` goes to Maven Central, finds the necessary .jar file, downloads it and unpacks into a given directory. You want to check that your class, that is using `Central` is asking it to download and unpack the right .jar. The best and and simplest approach would be to create `DummyCentral` class, which only creates an empty directory in the required path. Then, you just check the presence of such directory in your tests and that's it. 

What you are doing now with encapsulating mutable structures in MockMavenCentral and then checking their content afterwards is "asserting on side-effect" -- not a perfect object design.

volodya-lombrozo commented:
> @volodya-lombrozo please, forgive me for being picky. I think that `MockMavenCentral` is over-design in this case. Here is why: The live class `Central` goes to Maven Central, finds the necessary .jar file, downloads it and unpacks into a given directory. You want to check that your class, that is using `Central`, is asking it to download and unpack the right .jar. The best and and simplest approach would be to create `DummyCentral` class, which only creates an empty directory in the required path. Then, you just check the presence of such directory in your tests and that's it.
> 
> What you are doing now with encapsulating mutable structures in `MockMavenCentral` and then checking their content afterwards is "asserting on side-effect" -- not a perfect object design.

I've removed `MockMavenCentral` and added `DummyCenral` as you suggested. Please, check.

yegor256 commented on eo-maven-plugin/src/test/java/org/eolang/maven/LastPathSame.java:
```diff
@@ -0,0 +1,61 @@
+ *
+ * @since 0.28.11
+ */
+final class LastPathSame extends TypeSafeMatcher<Path> {
```
@volodya-lombrozo this class is redundant now, right?

volodya-lombrozo replied:
Yes, I've just forgot to remove the class. Done.

yegor256 commented on eo-maven-plugin/src/test/java/org/eolang/maven/LastDependencySame.java:
```diff
@@ -0,0 +1,107 @@
+ *
+ * @since 0.28.11
+ */
+final class LastDependencySame extends TypeSafeMatcher<Dependency> {
```
@volodya-lombrozo this class is not needed anymore, right?

volodya-lombrozo replied:
Yes, I've just forgot to remove the class. Done.

yegor256 commented:
@volodya-lombrozo check again, please. Looks like these two matchers are not needed anymore.

volodya-lombrozo commented:
@yegor256 can you merge the PR, please? If everything is fine, of course.

volodya-lombrozo commented:
## Updates for that PR

Sine we implemented `Dependencies` interface in that [PR](https://github.com/objectionary/eo/pull/1368). It’s convenient for that task to use the same approach with new classes:

`DcsWithRuntime` - adds Runtime dependency if it absents

`DcsWithoutConflicts` - checks conflicts (previously that class was a static method in `ResolveMojo`)

`DcsNoOneHasTransitive` - decorator to check all dependencies for transitive dependencies (previously it was a static method in `ResolveMojo`)


@mximp @yegor256  please, check and merge if you don't have any objections.

yegor256 commented on eo-maven-plugin/src/main/java/org/eolang/maven/DcsNoOneHasTransitive.java:
```diff
@@ -0,0 +1,119 @@
+        final Dependencies delegate,
+        final Func<Dependency, Dependencies> transitive
+    ) {
+        this.ignore = ignore;
```
@volodya-lombrozo read this, please: https://www.yegor256.com/2016/04/19/object-must-not-be-configurable.html

volodya-lombrozo replied:
@yegor256 fixed.

yegor256 commented on eo-maven-plugin/src/main/java/org/eolang/maven/DcsOf.java:
```diff
@@ -0,0 +1,55 @@
+ *
+ * @since 0.28.11
+ */
+final class DcsOf implements Dependencies {
```
@volodya-lombrozo what this class is for? Where and _why_ `Collection<Dependency>` is not good enough?

volodya-lombrozo replied:
I wanted to make it more obvious. We can't use it directly in the next construction:
```java
           new DcsNoOneHasTransitive(
                        this.ignoreTransitive,
                        dcs,
                        ....
           )
```
because `Collection<org.apache.maven.model.Dependency>` != `Dependencies`. But, of course, we can use it this way:
```java
           new DcsNoOneHasTransitive(
                        this.ignoreTransitive,
                        dcs::iterator,
                        ....
           )
```
I will fix it. Thank you.

volodya-lombrozo replied:
@yegor256 fixed.

yegor256 commented on eo-maven-plugin/src/main/java/org/eolang/maven/RuntimeDependencyEquality.java:
```diff
@@ -0,0 +1,69 @@
+import org.apache.maven.model.Dependency;
+
+/**
+ * RuntimeDependencyEquality is a class for checking if dependency is a runtime dependency.
```
@volodya-lombrozo are you sure this is an accurate description of functionality here? It looks wrong to me.

volodya-lombrozo replied:
@yegor256 fixed.

yegor256 commented on eo-maven-plugin/src/test/java/org/eolang/maven/DummyCentral.java:
```diff
@@ -0,0 +1,59 @@
+            );
+        } catch (final IOException ex) {
+            throw new IllegalStateException(
+                String.format("DummyCentral can't save dependency %s by path %s", dependency, path),
```
@volodya-lombrozo the information about the originator of the exception is in stacktrace. No need to add this noise. Just say `"Can't save '%s' to '%s'"`. In general, the shorter your messages, the better. They must be short, but contain maximum information.

yegor256 replied:
@volodya-lombrozo moreover, if I remove this class tomorrow, I most probably will forget to rename it in the text

volodya-lombrozo replied:
@yegor256 fixed

yegor256 commented on eo-maven-plugin/src/main/java/org/eolang/maven/DcsNoOneHasTransitive.java:
```diff
@@ -0,0 +1,105 @@
+    @SuppressWarnings("PMD.AvoidCatchingGenericException")
+    private boolean hasTransitive(final Dependency dep) {
+        try {
+            return new LengthOf(
```
@volodya-lombrozo use `Unchecked` from Cactoos, to avoid exception catching

volodya-lombrozo replied:
@yegor256 fixed.

yegor256 commented on eo-maven-plugin/src/main/java/org/eolang/maven/DcsWithoutConflicts.java:
```diff
@@ -0,0 +1,87 @@
+                )
+            );
+        if (!conflicts.isEmpty()) {
+            final String msg = String.format(
```
@volodya-lombrozo this is a bad practice: to do both logging and throwing. Pick one.

volodya-lombrozo replied:
@yegor256 it's your code :) I've just moved it to a separate class. Ok, I'll fix.

volodya-lombrozo replied:
@yegor256 I've chosen the exception. Done.

yegor256 commented on eo-maven-plugin/src/main/java/org/eolang/maven/ResolveMojo.java:
```diff
@@ -218,70 +213,53 @@ private Collection<Dependency> deps() throws IOException {
+    private Dependencies dependenciesOf(final Collection<Dependency> all) {
+        final Dependencies dependencies;
+        if (this.ignoreTransitive && this.ignoreVersionConflicts) {
+            dependencies = new DcsWithRuntime(all::iterator);
```
@volodya-lombrozo looks like code duplication to me. You use `DcsWithRuntime` everywhere in each `if`. Maybe it's better to make `dependencies` mutable?

volodya-lombrozo replied:
@yegor256 done.

yegor256 commented on eo-maven-plugin/src/test/java/org/eolang/maven/DummyCentral.java:
```diff
@@ -0,0 +1,59 @@
+            Files.createDirectories(path);
+            Files.createFile(
+                path.resolve(
+                    String.format("%s-%s.jar", dependency.getArtifactId(), dependency.getVersion())
```
@volodya-lombrozo you ignore `dependency.getClassifier()`, why? maybe it's better to use it too?

volodya-lombrozo replied:
@yegor256 added.

yegor256 commented:
@volodya-lombrozo a few more comments from me, see above

volodya-lombrozo commented:
@yegor256 I've applied all your suggestions above. Please, merge if you don't have any objections.

yegor256 commented:
@rultor merge

rultor commented:
> @rultor merge

@yegor256 OK, I'll try to merge now. You can check the progress of the merge [here](https://www.rultor.com/t/27774-1298352480)

rultor commented:
> @rultor merge

@yegor256 Done! FYI, the full log is [here](https://www.rultor.com/t/27774-1298352480) (took me 8min)
